### DISCUSSION TITLE
integrating the validator with Azure Databricks

### DISCUSSION THREAD

**Discussion by AEissa87** [CMSgov/price-transparency-guide-validator#80] [General]
Is there a way to integrate this validator with Azure Databricks? We are building MRFs using Databricks and SQL and we'd like to have them validated on the run within the same notebook without having to export then test them on premise.

**Comment by dgolden** [comment]
@AEissa87 I'm not savvy with Databricks, but it seems completely reasonable that you could spin up a VM in Databricks that would host the Validator. I have no idea how much effort would be required to make the process synchronous, to trigger the Validator to run immediately after the MRF was created.

Another approach is to contract out validation of your MRF to a third-party, at least for spot checking. The CMS Validator is a nice tool, checking the JSON schema for validity and doing some value range checks, but the data aggregators in this space see lots of issues with MRFs that the Validator was not designed to check. It's cheap insurance.

**Comment by shaselton-usds** [comment]
@AEissa87 

I'm inclined to agree with @dgolden -- this shouldn't be too hard to include the validator into your workflow. The validator should be completely modular. 

@dgolden what else would you want out of the validator beyond the schema checks that would broadly applicable to all MRFs? We are definitely open to suggestions. When it comes to validating the actual data, it gets hard (impossible?) if that data clears the schema rule thresholds.

**Comment by dgolden** [comment]
@shaselton-usds Thank you for enriching this thread. A couple suggestions come to mind:

a) Command line switch "--relaxed_schema_check=false" (default setting) which generates an error should the MRF contain any elements other than the CMS specified required and optional ones. We have encountered payers where an optional field was misspelled. MRF was broken, but the Validator didn't see it.

b) Support use of a URL for accessing a remote MRF. This avoids the cost of having a 1TB+ drive on the VM <grin>.  Or provide scripts (Unix and that Powershell) that seamlessly downloads an MRF and runs the Validator. From the GitHub discussion traffic my sense is that this task is outside the comfort level of many organizations, especially smaller consulting firms without dedicated IT staff.

**Reply by shaselton-usds** [reply]
Thanks for this!

A) We've looked into this a bit -- the rule itself allows for any attributes to be included outside of the required attributes (incase they help provide extra context that the producers of the files may want). This makes the validation checks of "optional" attributes a bit tricky. We'll continue to look into this...

B) We are pretty close to including support for URLs, although the behavior would be a bit different than what you've described initially in validating a remote file (not quite sure how to go about this if you have ideas....). The URL would automatically fetch the MRF and download it locally to a temp folder, run validation, produce results, and then delete the file.

**Comment by dgolden** [comment]
@shaselton-usds 
A) "attributes to be included outside of the required attributes" -> my 2 cents is that this is fraught with danger as: 
1) there is no control over different producers using the same attribute name but with different meanings
2) there is no reliable mechanism to let MRF consumers know that new attributes exist
3) unless the new attribute is of incredible value, nobody can justify customization of their code and DB
4) optional attributes leads to fragmentation of a standard.
The net-net is that CMS has to decide whether to have an MRF standard or not.  
It may be useful to look toward how the JSON Schema group defines their standards and evolves through voluntary industry cooperation. I can put you in touch with their leadership.

B) This solution is fine.

**Reply by shaselton-usds** [reply]
@dgolden Just flagging for your awareness, we created an issue #82 that has a PR #83 that address item "A"
